### PR TITLE
[SPARK-13407] Guard against garbage-collected accumulators in TaskMetrics.fromAccumulatorUpdates

### PR DESCRIPTION
**Pull request by JoshRosen**
`TaskMetrics.fromAccumulatorUpdates()` can fail if accumulators have been garbage-collected on the driver. To guard against this, this patch introduces `ListenerTaskMetrics`, a subclass of `TaskMetrics` which is used only in `TaskMetrics.fromAccumulatorUpdates()` and which eliminates the need to access the original accumulators on the driver.

### CODE REVIEW

**Comment by andrewor14** [comment]
LGTM

**Comment by JoshRosen** [comment]
Uh oh:

```
[info] - from accumulator updates *** FAILED *** (2 milliseconds)
[info]   0 did not equal 25 (TaskMetricsSuite.scala:539)
[info]   org.scalatest.exceptions.TestFailedException:
[info]   at org.scalatest.Assertions$class.newAssertionFailedException(Assertions.scala:500)
```

**Comment by JoshRosen** [comment]
Going to merge this into master.

File: core/src/main/scala/org/apache/spark/executor/TaskMetrics.scala [hunk 1]
```diff
@@ -364,6 +364,27 @@ class TaskMetrics private[spark] (initialAccums: Seq[Accumulator[_]]) extends Se
...
+ * Instances of this class should be considered read-only and users should not call `inc*()` or
+ * `set*()` methods. While we could override the setter methods to throw
+ * UnsupportedOperationException, we choose not to do so because the overrides would quickly become
+ * out-of-date when new metrics are added.
+ */
+private[spark] class ListenerTaskMetrics(
+    initialAccums: Seq[Accumulator[_]],
+    accumUpdates: Seq[AccumulableInfo]) extends TaskMetrics(initialAccums) {
+
+  override def accumulatorUpdates(): Seq[AccumulableInfo] = accumUpdates
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/executor/TaskMetrics.scala]: Does this PR avoid accessing the original accumulators on the driver only for `accumulatorUpdates`? Other methods like `resultSize` still reference to original accumulators right?

**Pull request closed without merging**